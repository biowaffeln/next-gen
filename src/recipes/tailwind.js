// @ts-check
import { outputFile, pathExists, writeFile } from "fs-extra"
import { withParser } from "jscodeshift"
import { updateApp, updatePackageJSON } from "../helpers/fs"
import { addImport } from "../helpers/jscodeshift"

/** @typedef {import("@/types/next-gen").Dependencies} Dependencies */

/** @type {Dependencies} */
const dependencies = {
	devDependencies: {
		autoprefixer: "^10.2.1",
		postcss: "^8.2.4",
		tailwindcss: "^2.0.2",
	},
}

const JS_TAILWIND_CONFIG = `module.exports = {
	purge: ["./pages/**/*.js", "./components/**/*.js"],
	darkMode: "media", // or "media" or "class"
	theme: {
		extend: {},
	},
	variants: {
		extend: {},
	},
	plugins: []
}`

const TS_TAILWIND_CONFIG = `module.exports = {
	purge: ["./components/**/*.{js,ts,jsx,tsx}", "./pages/**/*.{js,ts,jsx,tsx}"],
	darkMode: "media", // "media" or "class"
	theme: {
		extend: {}
	},
	variants: {
		extend: {},
	},
	plugins: [],
}`

const POSTCSS_CONFIG = `module.exports = {
  plugins: {
    tailwindcss: {},
    autoprefixer: {},
  },
}`

const STYLES = `@tailwind base;
@tailwind components;
@tailwind utilities;
`

/**
 * Adds Tailwind support to a Next.js project.
 * - installs dependencies
 * - adds tailwind and postcss config files
 * - adds an import to
 */
export async function recipeTailwind() {
	await updatePackageJSON(dependencies)
	if (await pathExists("tsconfig.json")) {
		writeFile("tailwind.config.js", TS_TAILWIND_CONFIG)
	} else {
		writeFile("tailwind.config.js", JS_TAILWIND_CONFIG)
	}
	await writeFile("postcss.config.js", POSTCSS_CONFIG)
	await outputFile("styles/tailwind.css", STYLES)
	await updateApp(addTailwindImport)
}

const j = withParser("tsx")

/**
 * @param {string} src
 * @returns {string}
 */
function addTailwindImport(src) {
	const root = j(src)
	addImport(j, root, `import "../styles/tailwind.css";`)
	return root.toSource()
}
