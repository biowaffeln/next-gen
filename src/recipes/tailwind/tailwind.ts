import { Recipe } from '../recipe';
import { addImport } from '../transforms/addImport';
import { getAppPath } from '../pathHelpers/getPagePath';
import { readTemplatePath } from '../pathHelpers/readTemplatePath';
import * as j from 'jscodeshift';

const getTemplate = readTemplatePath(__dirname);

export const tailwind: Recipe = async () => ({
	addDependencies: [
		{ package: 'tailwindcss', version: 'latest', dev: true },
		{ package: 'autoprefixer', version: 'latest', dev: true },
		{ package: 'postcss', version: 'latest', dev: true },
	],
	createFiles: [
		{
			path: 'tailwind.config.js',
			content: await getTemplate('tailwind.config.template'),
			isRoot: true,
		},
		{
			path: 'postcss.config.js',
			content: await getTemplate('postcss.config.template'),
			isRoot: true,
		},
		{
			path: 'styles/global.css',
			content: await getTemplate('global.template'),
		},
	],
	ASTTransforms: [
		{
			path: getAppPath(),
			transform: (root) =>
				addImport(root, j.template.statement`import "../styles/global.css"`),
			fallback: '',
			parser: 'tsx',
		},
	],
});
