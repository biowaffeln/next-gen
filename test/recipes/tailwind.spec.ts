import { tailwind } from '../../src/recipes/tailwind/tailwind';

const packageJSON = `{
	"name": "test-app",
	"version": "1.0.0",
	"dependencies": {
		"react": "latest",
		"react-dom": "latest",
		"next": "latest"
	}
}`;

const appJS = `export default function App({ Component, pageProps }) {
	return (
		<Component {...pageProps} />
	);
}
`;

test('should transform files correctly', () => {
	const [pkg, styles, app] = tailwind();
	expect(pkg.transform(packageJSON)).toMatchSnapshot();
	expect(app.transform(appJS)).toMatchSnapshot();
	expect(styles.transform('')).toMatchSnapshot();
});
