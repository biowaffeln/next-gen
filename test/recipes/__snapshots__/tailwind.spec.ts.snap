// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should transform files correctly 1`] = `
"{
  \\"name\\": \\"test-app\\",
  \\"version\\": \\"1.0.0\\",
  \\"dependencies\\": {
    \\"react\\": \\"latest\\",
    \\"react-dom\\": \\"latest\\",
    \\"next\\": \\"latest\\"
  },
  \\"devDependencies\\": {
    \\"tailwindcss\\": \\"^3.0.0\\",
    \\"postcss\\": \\"^8.4.4\\",
    \\"autoprefixer\\": \\"^10.4.0\\"
  }
}"
`;

exports[`should transform files correctly 2`] = `
"import \\"../styles/global.css\\";
 
export default function App({ Component, pageProps }) {
	return (
		<Component {...pageProps} />
	);
}
"
`;

exports[`should transform files correctly 3`] = `
"@tailwind base;
@tailwind components;
@tailwind utilities;
"
`;
